solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,380 @@
+        pendingOptimisticMints[depositKey] = block.timestamp;
+
+        emit OptimisticMintingRequested(
+            msg.sender,
+            deposit.depositor,
+            deposit.amount,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a Minter to finalize previously requested optimistic
+    ///         minting. The following conditions must be met:
+    ///         - The optimistic minting has been requested for the given
+    ///           deposit.
+    ///         - The deposit has not been swept yet.
+    ///         - At least OPTIMISTIC_MINTING_DELAY passed since the optimistic
+    ///           minting was requested for the given deposit.
+    ///         - The optimistic minting has not been finalized earlier for the
+    ///           given deposit.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been canceled by a Guardian.
+    ///         - The optimistic minting is not paused.
+    ///         This function mints TBTC and increases pendingOptimisticMints
+    ///         for the given depositor. The finalized optimistic minting
+    ///         request is removed from the contract.
+    function finalizeOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyMinter whenOptimisticMintingNotPaused {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        uint256 requestedAt = pendingOptimisticMints[depositKey];
+        require(
+            requestedAt != 0,
+            "Optimistic minting not requested or already finalized"
+        );
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp - requestedAt > OPTIMISTIC_MINTING_DELAY,
+            "Optimistic minting delay has not passed yet"
+        );
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        require(deposit.sweptAt == 0, "The deposit is already swept");
+
+        // Bridge, when sweeping, cuts a deposit treasury fee and splits
+        // Bitcoin miner fee for the sweep transaction evenly between the
+        // depositors in the sweep.
+        //
+        // When tokens are optimistically minted, we do not know what the
+        // Bitcoin miner fee for the sweep transaction will look like.
+        // The Bitcoin miner fee is ignored. When sweeping, the miner fee is
+        // subtracted so the optimisticMintingDebt may stay non-zero after the
+        // deposit is swept.
+        //
+        // This imbalance is supposed to be solved by a donation to the Bridge.
+        uint256 amountToMint = deposit.amount - deposit.treasuryFee;
+        _mint(deposit.depositor, amountToMint);
+        optimisticMintingDebt[deposit.depositor] += amountToMint;
+
+        delete pendingOptimisticMints[depositKey];
```

Comment:
Deleting can be problematic as the same deposit key can be reused for optimistic minting if the deposit was not yet swept. If we decide to not delete, we should also prevent `pendingOptimisticMints` duplicates in the `optimisticMint` function. Unless we want to push that responsibility to the guardians? If so, we should clearly note that requirement.